cpp/oneapi/dal/detail/common.hpp

```diff
@@ -166,4 +169,26 @@ inline void check_mul_overflow(const Data& first, const Data& second) {
     integer_overflow_ops<Data>{}.check_mul_overflow(first, second);
 }
 
+template <typename Data>
+struct limits {
+    static constexpr Data min() {
+        return std::numeric_limits<Data>::min();
+    }
+    static constexpr Data max() {
+        return std::numeric_limits<Data>::max();
+    }
+};
+
+template <typename Out, typename In>
+inline Out integral_cast(const In& value) {
```

Comment:
I see several problems with that function in general case
- [Risk of signed vs unsigned comparison](https://softwareengineering.stackexchange.com/questions/175253/why-does-an-unsigned-int-compared-with-a-signed-character-turn-out-with-an-unexp) here `value <= (limits<Out>::max()` and here `value >= (limits<Out>::min())`
- In some cases assertions are not enough and errors must be reported to user, we cannot detect whether error should be reported or not within the function, so it should be decision of the function caller.

How to deal with it:
- Split this function into four functions: signed-signed, signed-unsigned, unsigned-signed, unsigned-unsigned. Use `std::enable_if_t` to distinguish between different cases. Another option keep one function and use `if constexpr`, maybe it's even better.
- Change name of the function to `integral_asserted_cast` in the future it might be another function called `integral_checked_cast` that would throw exception.